lib/cn/cn.c

```diff
@@ -941,57 +951,64 @@ cn_open(
-            if (tn->tn_nodeid > 0) {
-                cn_tree_node_get_max_key(tn, kbuf, sizeof(kbuf), &klen);
-
-                log_info("nodeid %lu, klen %u", tn->tn_nodeid, klen);
+    /* Walk the list of nodes created/populated by cn_kvset_cb() and insert
+     * all leaf nodes into the route map (i.e., all nodes except the root
+     * node which always has node ID 0).
+     */
+    list_for_each_entry(tn, &cn->cn_tree->ct_nodes, tn_link) {
+        if (tn->tn_nodeid > 0) {
+            cn_tree_node_get_max_key(tn, kbuf, sizeof(kbuf), &klen);
 
-                tn->tn_route_node = route_map_insert(cn->cn_tree->ct_route_map, tn, kbuf, klen);
-                if (!tn->tn_route_node) {
-                    err = merr(EINVAL);
-                    goto err_exit;
-                }
+            tn->tn_route_node = route_map_insert(cn->cn_tree->ct_route_map, tn, kbuf, klen);
+            if (!tn->tn_route_node) {
+                err = merr(EINVAL);
+                goto err_exit;
             }
         }
+    }
 
-        klen = sizeof(kbuf);
-        memset(kbuf, -1, klen);
-
-        rn = route_map_lookup(cn->cn_tree->ct_route_map, kbuf, klen);
-        if (rn) {
-            tn = route_node_tnode(rn);
-            if (!tn)
-                abort();
-
-            log_info("updating node %lu right edge key", tn->tn_nodeid);
+    /* Initialize kbuf to the largest possible key.
+     */
+    klen = sizeof(kbuf);
+    memset(kbuf, -1, klen);
 
-            err = route_node_key_modify(cn->cn_tree->ct_route_map, rn, kbuf, klen);
-            if (err)
-                goto err_exit;
-        } else {
-            tn = cn_node_alloc(cn->cn_tree, cndb_nodeid_mint(cndb));
-            if (!tn) {
-                err = merr(ENOMEM);
-                goto err_exit;
-            }
+    /* Find the node in the route map with the largest edge key and force
+     * it to the the largest possible key.  If the route map is empty then
+     * create the initial level-2 node with the largest possible edge key.
+     */
+    rn = route_map_lookup(cn->cn_tree->ct_route_map, kbuf, klen);
+    if (rn) {
+        tn = route_node_tnode(rn);
+        if (!tn)
+            abort();
+
+        err = route_node_key_modify(cn->cn_tree->ct_route_map, rn, kbuf, klen);
+        if (err)
+            goto err_exit;
+    } else {
+        tn = cn_node_alloc(cn->cn_tree, cndb_nodeid_mint(cndb));
+        if (!tn) {
+            err = merr(ENOMEM);
+            goto err_exit;
+        }
 
-            log_info("creating node %lu right edge key", tn->tn_nodeid);
+        list_add_tail(&tn->tn_link, &cn->cn_tree->ct_nodes);
 
-            list_add_tail(&tn->tn_link, &cn->cn_tree->ct_nodes);
-            tn->tn_route_node = route_map_insert(cn->cn_tree->ct_route_map, tn, kbuf, klen);
+        tn->tn_route_node = route_map_insert(cn->cn_tree->ct_route_map, tn, kbuf, klen);
+        if (!tn->tn_route_node) {
+            err = merr(ENOMEM);
+            goto err_exit;
```

Comment:
Crap, I left a leve-2 in there..  Nabeel, would you mind fixing that comment?

Reply:
Done, s/level-2/leaf